cases/src/org/commcare/cases/util/CasePurgeFilter.java

```diff
@@ -222,7 +227,96 @@ private void propogateMarkToDAG(DAG<String, int[], String> dag, boolean walkFrom
+     * phone for some reason). Then remove any nodes that are made invalid by that parent node
+     * not existing
+     *
+     * @return Whether or not this method invocation removed any invalid edges from the DAG
+     */
+    private boolean removeInvalidEdges() {
+        Hashtable<String, Vector<Edge<String, String>>> allEdges = internalCaseDAG.getEdges();
+        Set<String> childOfNonexistentParent = new HashSet<>();
+        Set<String[]> edgesToRemove = new HashSet<>();
+        boolean invalidEdgesRemoved = false;
+        for (String originIndex : allEdges.keySet()) {
+            Vector<Edge<String, String>> edgeListForOrigin = allEdges.get(originIndex);
+            for (Edge<String, String> edge : edgeListForOrigin) {
+                String targetIndex = edge.i;
+                if (internalCaseDAG.getNode(targetIndex) == null) {
+                    invalidEdgesRemoved = true;
+                    edgesToRemove.add(new String[]{originIndex, targetIndex});
+                    childOfNonexistentParent.add(originIndex);
+                }
+            }
+        }
+
+        // Any case node with a nonexistent parent should be removed from the graph, as should
+        // all of its descendants, and any edges to or from it
+        for (String index : childOfNonexistentParent) {
+            edgesToRemove = removeNodeAndPropagate(index, edgesToRemove);
+        }
+
+        // Do actual edge removal after we finish all iterations through the edge list, to prevent
+        // ConcurrentModificationException
+        for (String[] edge : edgesToRemove) {
+            internalCaseDAG.removeEdge(edge[0], edge[1]);
+        }
+
+        return invalidEdgesRemoved;
+    }
+
+    /**
+     * Remove from the graph the node at this index, and remove all nodes that are made
+     * invalid by the non-existence of that node (i.e. all of its child and extension cases)
+     *
+     * @return A list of edges that need to be removed as a result of this method call. Actual edge
+     * removal should then be done all at once by the caller, to avoid
+     * ConcurrentModificationException
+     */
+    private Set<String[]> removeNodeAndPropagate(String indexOfRemovedNode,
+                                        Set<String[]> accumulatedEdgesToRemove) {
+        // Wording is confusing here -- because all edges in this graph are from a child case to
+        // a parent case, calling getParents() for a node returns all of its child/extension cases
+        Vector<Edge<String, String>> childCases = internalCaseDAG.getParents(indexOfRemovedNode);
+        for (Edge<String, String> child : childCases) {
+            // Want to remove the edge from child case to parent case
+            accumulatedEdgesToRemove.add(new String[]{child.i, indexOfRemovedNode});
+
+            // Recurse on child case
+            accumulatedEdgesToRemove = removeNodeAndPropagate(child.i, accumulatedEdgesToRemove);
```

Comment:
This assignment shouldn't be needed, right?
just `removeNodeAndPropagate(child.i, accumulatedEdgesToRemove);`